Common/Python/PythonInitializer.cs

```diff
@@ -57,31 +60,82 @@ public static void Initialize()
         /// <summary>
         /// Adds directories to the python path at runtime
         /// </summary>
-        public static void AddPythonPaths(IEnumerable<string> paths)
+        public static void AddPythonPaths(IEnumerable<string> paths, bool prepend = false)
         {
-            if (paths == null)
+            // Filter out any paths that are already on our Python path
+            paths = paths?.Where(x => !pathCache.Contains(x.Replace('\\', '/'))).ToList();
+            if (paths == null || !paths.Any())
             {
                 return;
             }
 
+            // Add these paths to our pending additions
+            if (prepend)
+            {
+                _pendingStartOfPathAdditions.AddRange(paths);
+            }
+            else
+            {
+                _pendingEndOfPathAdditions.AddRange(paths);
+            }
+
             if (_isInitialized)
             {
                 using (Py.GIL())
                 {
-                    _pendingPathAdditions.AddRange(paths);
+                    dynamic sys = Py.Import("sys");
+                    var locals = new PyDict();
+                    locals.SetItem("sys", sys);
+
+                    // Insert any pending start of path additions
+                    if (!_pendingStartOfPathAdditions.IsNullOrEmpty())
+                    {
+                        PythonEngine.Exec(string.Join(";", _pendingStartOfPathAdditions.Select(s => $"sys.path.insert(0, '{s}')"))
```

Comment:
Think using `;` or `:` would depend on the OS?

Reply:
Not applicable here because we are just modifying Pythons sys.path directly with appending/insert calls. The `;` is just to separate those calls into individual python statements in one line.